File: backend/app/schemas/ai.py
from datetime import datetime
from typing import Any
from pydantic import BaseModel, Field


class TaskBreakdownRequest(BaseModel):
    title: str
    description: str | None = None
    priority: str | None = None


class SubtaskSuggestion(BaseModel):
    title: str
    estimated_hours: float = 1.0


class TaskBreakdownResponse(BaseModel):
    suggested_description: str
    suggested_priority: str
    suggested_tags: list[str]
    subtasks: list[SubtaskSuggestion]


class SprintAnalysisResponse(BaseModel):
    health_score: int
    health_status: str
    total_tasks: int
    completed_tasks: int
    overdue_tasks: int
    predicted_blockers: list[str]
    recommendations: list[str]


class DailyStandupResponse(BaseModel):
    generated_at: datetime
    workspace_name: str
    completed_recent: list[str]
    in_progress_today: list[str]
    blockers_and_risks: list[str]
    summary_markdown: str


class SprintRetrospectiveResponse(BaseModel):
    generated_at: datetime
    workspace_name: str
    what_went_well: list[str]
    what_could_be_improved: list[str]
    action_items: list[str]
    velocity_summary: dict[str, Any]
    summary_markdown: str


class AutoAssignRequest(BaseModel):
    task_title: str
    task_priority: str = "medium"


class AutoAssignResponse(BaseModel):
    recommended_user_id: str | None
    recommended_name: str | None
    current_active_tasks: int
    reason: str


class KnowledgeDocCreate(BaseModel):
    title: str
    content: str
    tags: list[str] = Field(default_factory=list)


class KnowledgeDocResponse(BaseModel):
    id: str
    workspace_id: str
    title: str
    content: str
    tags: list[str]
    created_at: datetime


class KnowledgeSearchRequest(BaseModel):
    query: str


class KnowledgeSearchResponse(BaseModel):
    results: list[dict[str, Any]]
    answer: str


class ChatMessageItem(BaseModel):
    role: str
    content: str


class ExternalAIChatRequest(BaseModel):
    prompt: str
    provider: str | None = "openai"  # openai, groq, openrouter, custom
    api_key: str | None = None
    model: str | None = None
    endpoint: str | None = None
    history: list[ChatMessageItem] = Field(default_factory=list)
    file_name: str | None = None
    file_type: str | None = None
    file_data: str | None = None  # Base64 string for photo/file upload



class ExternalAIChatResponse(BaseModel):
    reply: str
    model_used: str
    provider: str
    suggested_action: dict[str, str] | None = None

File: frontend/src/services/ai.ts
import { api } from "./api";

export interface SubtaskSuggestion {
  title: string;
  estimated_hours: number;
}

export interface TaskBreakdownResponse {
  suggested_description: string;
  suggested_priority: string;
  suggested_tags: string[];
  subtasks: SubtaskSuggestion[];
}

export interface SprintAnalysisResponse {
  health_score: number;
  health_status: string;
  total_tasks: number;
  completed_tasks: number;
  overdue_tasks: number;
  predicted_blockers: string[];
  recommendations: string[];
}

export interface DailyStandupResponse {
  generated_at: string;
  workspace_name: string;
  completed_recent: string[];
  in_progress_today: string[];
  blockers_and_risks: string[];
  summary_markdown: string;
}

export interface AutoAssignResponse {
  recommended_user_id: string | null;
  recommended_name: string | null;
  current_active_tasks: number;
  reason: string;
}

export interface KnowledgeSearchResponse {
  results: {
    type: "project" | "task" | "document";
    id: string;
    title: string;
    snippet: string;
  }[];
  answer: string;
}

export interface ExternalAIChatRequest {
  prompt: string;
  provider?: string;
  api_key?: string;
  model?: string;
  endpoint?: string;
  history?: { role: string; content: string }[];
  file_name?: string;
  file_type?: string;
  file_data?: string;
}

export interface ExternalAIChatResponse {
  reply: string;
  model_used: string;
  provider: string;
  suggested_action?: {
    label: string;
    path: string;
  };
}

/* ============================================================
   AI API METHODS
============================================================ */

export async function chatAI(
  workspaceId: string,
  data: ExternalAIChatRequest,
): Promise<ExternalAIChatResponse> {
  const response = await api.post<ExternalAIChatResponse>(
    `/api/v1/workspaces/${workspaceId}/ai/chat`,
    data,
  );
  return response.data;
}

export async function getAITaskBreakdown(
  workspaceId: string,
  title: string,
  description?: string,
  priority?: string,
): Promise<TaskBreakdownResponse> {
  const response = await api.post<TaskBreakdownResponse>(
    `/api/v1/workspaces/${workspaceId}/ai/task-breakdown`,
    { title, description, priority },
  );
  return response.data;
}

export async function getAISprintAnalysis(
  workspaceId: string,
): Promise<SprintAnalysisResponse> {
  const response = await api.get<SprintAnalysisResponse>(
    `/api/v1/workspaces/${workspaceId}/ai/sprint-analysis`,
  );
  return response.data;
}

export async function getAIDailyStandup(
  workspaceId: string,
): Promise<DailyStandupResponse> {
  const response = await api.get<DailyStandupResponse>(
    `/api/v1/workspaces/${workspaceId}/ai/daily-standup`,
  );
  return response.data;
}

export async function getAIAutoAssign(
  workspaceId: string,
  taskTitle: string,
  taskPriority: string = "medium",
): Promise<AutoAssignResponse> {
  const response = await api.post<AutoAssignResponse>(
    `/api/v1/workspaces/${workspaceId}/ai/auto-assign`,
    { task_title: taskTitle, task_priority: taskPriority },
  );
  return response.data;
}

export async function searchAIKnowledge(
  workspaceId: string,
  query: string,
): Promise<KnowledgeSearchResponse> {
  const response = await api.post<KnowledgeSearchResponse>(
    `/api/v1/workspaces/${workspaceId}/ai/knowledge-search`,
    { query },
  );
  return response.data;
}
